File: dotnet/MaigoLabs.NeedLe.Indexer/Tokenizer.cs
using MaigoLabs.NeedLe.Common;
using MaigoLabs.NeedLe.Common.Extensions;
using MaigoLabs.NeedLe.Common.Types;
using MaigoLabs.NeedLe.Indexer.Han;
using MaigoLabs.NeedLe.Indexer.Japanese;

namespace MaigoLabs.NeedLe.Indexer;

public class TokenizerOptions
{
    public HanVariantProvider? HanVariantProvider { get; set; }
    public TranscriptionProvider? TranscriptionProvider { get; set; }
}

public class Tokenizer(TokenizerOptions? options = null)
{
    public HanVariantProvider HanVariantProvider { get; set; } = options?.HanVariantProvider ?? new HanVariantProvider();
    public TranscriptionProvider TranscriptionProvider { get; set; } = options?.TranscriptionProvider ?? new TranscriptionProvider();

    public class Token
    {
        public required int Id { get; set; }
        public required int Start { get; set; }
        public required int End { get; set; }
    }

    public Dictionary<(TokenType Type, string Text), TokenDefinition> Tokens { get; } = [];
    private TokenDefinition EnsureToken(TokenType type, string text)
    {
        var key = (type, text);
        if (Tokens.TryGetValue(key, out var tokenDefinition)) return tokenDefinition;
        tokenDefinition = new TokenDefinition { Id = Tokens.Count, Type = type, Text = text, CodePointLength = text.ToCodePoints().Count() };
        Tokens.Add(key, tokenDefinition);
        return tokenDefinition;
    }

    public List<Token> Tokenize(string text)
    {
        var codePoints = text.ToCodePoints().Select(CommonNormalization.NormalizeCodePoint).ToArray();
        var results = new List<Token>();
        Action<TokenType /* tokenType */, string /* text */> Emitter(int start, int end) =>
            (tokenType, codePoints) => results.Add(new Token { Id = EnsureToken(tokenType, codePoints).Id, Start = start, End = end });

        void EmitMaybeJapanese(ReadOnlyMemory<int> codePoints, int offset)
        {
            foreach (var combination in TranscriptionProvider.EnumerateKanaTranscriptions(codePoints))
            {
                var emit = Emitter(offset + combination.Start, offset + combination.Start + combination.Length);
                foreach (var transcription in combination.Transcriptions) emit(TokenType.Kana, transcription);
            }
            foreach (var combination in TranscriptionProvider.EnumerateRomajiTranscriptions(codePoints))
            {
                var emit = Emitter(offset + combination.Start, offset + combination.Start + combination.Length);
                foreach (var transcription in combination.Transcriptions) emit(TokenType.Romaji, transcription);
            }
            for (int i = 0; i < codePoints.Length; i++)
            {
                // Single character may have not only kana readings, but also Chinese pronunciations or Simplified/Traditional/Japanese variants.
                var hanAlternates = HanVariantProvider.GetHanVariants(codePoints.Span[i]); // All possible variant characters (Simplified/Traditional/Japanese)
                var pinyinAlternates = hanAlternates.SelectMany(PinyinHelper.GetPinyinCandidates).Distinct();
                var emit = Emitter(offset + i, offset + i + 1);
                foreach (var han in hanAlternates) emit(TokenType.Han, char.ConvertFromUtf32(han));
                foreach (var pinyin in pinyinAlternates) emit(TokenType.Pinyin, pinyin);
            }
        }

        var consequentCharsets = new (Func<int, bool> Is, Action<ReadOnlyMemory<int>, int> Emit)[]
        {
            (Is: JapaneseUtils.IsMaybeJapanese, Emit: EmitMaybeJapanese),
        };

        void EmitRaw(int codePoint, int offset) => Emitter(offset, offset + 1)(TokenType.Raw, char.ConvertFromUtf32(codePoint));

        for (int start = 0; start < codePoints.Length; )
        {
            var codePoint = codePoints[start];
            var emitted = false;
            foreach (var (Is, Emit) in consequentCharsets)
            {
                var length = 0;
                while (start + length < codePoints.Length && Is(codePoints[start + length])) length++;
                if (length > 0)
                {
                    Emit(new Memory<int>(codePoints, start, length), start);
                    start += length;
                    emitted = true;
                    break;
                }
            }
            if (emitted) continue;

            // Skip whitespaces
            if (CommonUtils.IsWhitespace(codePoint))
            {
                start++;
                continue;
            }

            EmitRaw(codePoint, start);
            start++;
        }
        return results;
    }
}

File: packages/needle/src/indexer/tokenizer.ts
import type { TokenizerBuilder } from '@patdx/kuromoji';

import { getHanVariants, getPinyinCandidates } from './han';
import { createKanaTranscriptionEnumerator, createRomajiTranscriptionEnumerator, isMaybeJapanese } from './japanese';
import { normalizeByCodePoint } from '../common/normalize';
import { TokenType, type TokenDefinition } from '../common/types';

export interface Token {
  id: number;
  start: number;
  end: number;
}

export type KuromojiTokenizer = Awaited<ReturnType<TokenizerBuilder['build']>>;
export interface TokenizerOptions {
  kuromoji: KuromojiTokenizer;
}
export const createTokenizer = (options: TokenizerOptions) => {
  const tokens = new Map<string, TokenDefinition>();
  let nextId = 0;
  const ensureToken = (type: TokenType, text: string) => {
    const key = `${type}:${text}`;
    let tokenDefinition = tokens.get(key);
    if (tokenDefinition) return tokenDefinition;
    tokenDefinition = { id: nextId++, type, text, codePointLength: [...text].length };
    tokens.set(key, tokenDefinition);
    return tokenDefinition;
  };

  const enumerateAllKanaCombinations = createKanaTranscriptionEnumerator(options.kuromoji);
  const enumerateAllRomajiCombinations = createRomajiTranscriptionEnumerator(options.kuromoji);
  const tokenize = (text: string) => {
    const results: Token[] = [];
    const emitter = (start: number, end: number) => (type: TokenType, text: string) => results.push({ id: ensureToken(type, text).id, start, end });

    const emitMaybeJapanese = (codePoints: string[], offset: number) => {
      for (const { start, length, transcriptions } of enumerateAllKanaCombinations(codePoints)) {
        const emit = emitter(offset + start, offset + start + length);
        for (const transcription of transcriptions) emit(TokenType.Kana, transcription);
      }
      for (const { start, length, transcriptions } of enumerateAllRomajiCombinations(codePoints)) {
        const emit = emitter(offset + start, offset + start + length);
        for (const transcription of transcriptions) emit(TokenType.Romaji, transcription);
      }
      for (let i = 0; i < codePoints.length; i++) {
        // Single character may have not only kana readings, but also Chinese pronunciations or Simplified/Traditional/Japanese variants.
        const character = codePoints[i]!;
        const hanAlternates = getHanVariants(character); // All possible variant characters (Simplified/Traditional/Japanese)
        const pinyinAlternates = Array.from(new Set(hanAlternates.flatMap(han => getPinyinCandidates(han)))); // All possible pinyin candidates
        const emit = emitter(offset + i, offset + i + 1);
        for (const han of hanAlternates) emit(TokenType.Han, han);
        for (const pinyin of pinyinAlternates) emit(TokenType.Pinyin, pinyin);
      }
    };
    const emitRaw = (codePoint: string, offset: number) => emitter(offset, offset + 1)(TokenType.Raw, codePoint);

    const codePoints = [...normalizeByCodePoint(text)];
    for (let start = 0; start < codePoints.length;) {
      const codePoint = codePoints[start]!;

      const consequentCharsets = [
        { is: isMaybeJapanese, emit: emitMaybeJapanese },
      ];
      let emitted = false;
      for (const { is, emit } of consequentCharsets) {
        let length = 0;
        while (start + length < codePoints.length && is(codePoints[start + length]!)) length++;
        if (length > 0) {
          emit(codePoints.slice(start, start + length), start);
          start += length;
          emitted = true;
          break;
        }
      }
      if (emitted) continue;

      // Skip whitespaces
      if (/\s/.test(codePoint)) {
        start++;
        continue;
      }

      emitRaw(codePoint, start);
      start++;
    }
    return results;
  };

  return {
    tokens,
    tokenize,
  };
};
